File: sdk/dart/lib/src/collection.dart
import 'dart:convert';
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import 'bindings.dart' as lib;
import 'filter.dart';
import 'libravdb.dart';

class Collection {
  final int _handle;
  final int _dim;

  Collection(this._handle, this._dim);

  void _checkError(Pointer<Utf8> errPtr, String opName) {
    if (errPtr != nullptr) {
      final msg = errPtr.toDartString();
      lib.freeString(errPtr);
      if (msg.startsWith('error: ')) {
        throw LibraException('$opName failed: ${msg.substring(7)}');
      }
      throw LibraException('$opName failed: $msg');
    }
  }

  dynamic _parseResult(Pointer<Utf8> resPtr, String opName) {
    if (resPtr == nullptr) {
      return null;
    }
    final msg = resPtr.toDartString();
    lib.freeString(resPtr);

    if (msg.isEmpty) {
      return null;
    }

    try {
      final node = jsonDecode(msg);
      if (node is Map && node.containsKey('error')) {
        throw LibraException('$opName failed: ${node['error']}');
      }
      return node;
    } catch (e) {
      if (e is LibraException) rethrow;
      throw LibraException('JSON Parse error: $e');
    }
  }

  void insert(String id, List<double> vector, {Map<String, dynamic>? metadata}) {
    if (vector.length != _dim) throw LibraException('Vector dimension mismatch');
    
    final idPtr = id.toNativeUtf8();
    final metaStr = metadata != null ? jsonEncode(metadata) : '';
    final metaPtr = metaStr.toNativeUtf8();
    
    final vecPtr = malloc.allocate<Float>(_dim * sizeOf<Float>());
    for (int i = 0; i < _dim; i++) {
      vecPtr[i] = vector[i];
    }

    final errPtr = lib.insertVector(_handle, idPtr, vecPtr, _dim, metaPtr);
    
    malloc.free(vecPtr);
    malloc.free(idPtr);
    malloc.free(metaPtr);
    
    _checkError(errPtr, 'Insert');
  }

  void update(String id, List<double> vector, {Map<String, dynamic>? metadata}) {
    if (vector.length != _dim) throw LibraException('Vector dimension mismatch');
    
    final idPtr = id.toNativeUtf8();
    final metaStr = metadata != null ? jsonEncode(metadata) : '';
    final metaPtr = metaStr.toNativeUtf8();
    
    final vecPtr = malloc.allocate<Float>(_dim * sizeOf<Float>());
    for (int i = 0; i < _dim; i++) {
      vecPtr[i] = vector[i];
    }

    final errPtr = lib.updateVector(_handle, idPtr, vecPtr, _dim, metaPtr);
    
    malloc.free(vecPtr);
    malloc.free(idPtr);
    malloc.free(metaPtr);
    
    _checkError(errPtr, 'Update');
  }

  void updateIfVersion(String id, List<double> vector, int expectedVersion, {Map<String, dynamic>? metadata}) {
    if (vector.length != _dim) throw LibraException('Vector dimension mismatch');
    
    final idPtr = id.toNativeUtf8();
    final metaStr = metadata != null ? jsonEncode(metadata) : '';
    final metaPtr = metaStr.toNativeUtf8();
    
    final vecPtr = malloc.allocate<Float>(_dim * sizeOf<Float>());
    for (int i = 0; i < _dim; i++) {
      vecPtr[i] = vector[i];
    }

    final errPtr = lib.updateVectorIfVersion(_handle, idPtr, vecPtr, _dim, metaPtr, expectedVersion);
    
    malloc.free(vecPtr);
    malloc.free(idPtr);
    malloc.free(metaPtr);
    
    _checkError(errPtr, 'UpdateIfVersion');
  }

  dynamic get(String id) {
    final idPtr = id.toNativeUtf8();
    final resPtr = lib.getVector(_handle, idPtr);
    malloc.free(idPtr);
    return _parseResult(resPtr, 'Get');
  }

  dynamic search(List<double> vector, int k, {Filter? filter}) {
    if (vector.length != _dim) throw LibraException('Vector dimension mismatch');
    
    final filterStr = filter != null ? filter.toJsonString() : '';
    final filterPtr = filterStr.toNativeUtf8();
    
    final vecPtr = malloc.allocate<Float>(_dim * sizeOf<Float>());
    for (int i = 0; i < _dim; i++) {
      vecPtr[i] = vector[i];
    }

    final resPtr = lib.queryVector(_handle, vecPtr, _dim, k, filterPtr);
    
    malloc.free(vecPtr);
    malloc.free(filterPtr);
    
    return _parseResult(resPtr, 'Search');
  }

  dynamic scan(int offset, int limit) {
    final resPtr = lib.scanCollection(_handle, offset, limit);
    return _parseResult(resPtr, 'Scan');
  }

  void insertBatch(List<String> ids, List<List<double>> vectors, {List<Map<String, dynamic>>? metadata}) {
    int count = ids.size;
    if (vectors.length != count) throw LibraException('ids and vectors size mismatch');

    final idsPtr = malloc.allocate<Pointer<Utf8>>(count * sizeOf<Pointer<Utf8>>());
    for (int i = 0; i < count; i++) {
      idsPtr[i] = ids[i].toNativeUtf8();
    }

    final vecsPtr = malloc.allocate<Float>(count * _dim * sizeOf<Float>());
    for (int i = 0; i < count; i++) {
      if (vectors[i].length != _dim) throw LibraException('Vector dimension mismatch at index $i');
      for (int j = 0; j < _dim; j++) {
        vecsPtr[i * _dim + j] = vectors[i][j];
      }
    }

    Pointer<Pointer<Utf8>> metasPtr = nullptr;
    if (metadata != null) {
      if (metadata.length != count) throw LibraException('ids and metadata size mismatch');
      metasPtr = malloc.allocate<Pointer<Utf8>>(count * sizeOf<Pointer<Utf8>>());
      for (int i = 0; i < count; i++) {
        metasPtr[i] = jsonEncode(metadata[i]).toNativeUtf8();
      }
    }

    final errPtr = lib.insertBatch(_handle, idsPtr, vecsPtr, count, _dim, metasPtr);

    // Free everything
    for (int i = 0; i < count; i++) {
      malloc.free(idsPtr[i]);
      if (metasPtr != nullptr) {
        malloc.free(metasPtr[i]);
      }
    }
    malloc.free(idsPtr);
    malloc.free(vecsPtr);
    if (metasPtr != nullptr) {
      malloc.free(metasPtr);
    }

    _checkError(errPtr, 'InsertBatch');
  }

  void deleteBatch(List<String> ids) {
    int count = ids.length;
    final idsPtr = malloc.allocate<Pointer<Utf8>>(count * sizeOf<Pointer<Utf8>>());
    for (int i = 0; i < count; i++) {
      idsPtr[i] = ids[i].toNativeUtf8();
    }

    final errPtr = lib.deleteBatch(_handle, idsPtr, count);

    for (int i = 0; i < count; i++) {
      malloc.free(idsPtr[i]);
    }
    malloc.free(idsPtr);

    _checkError(errPtr, 'DeleteBatch');
  }

  int count() {
    final c = lib.getCollectionCount(_handle);
    if (c < 0) throw LibraException('Failed to get collection count');
    return c;
  }

  void enableMemoryMapping(String path) {
    final pathPtr = path.toNativeUtf8();
    final errPtr = lib.enableMemoryMapping(_handle, pathPtr);
    malloc.free(pathPtr);
    _checkError(errPtr, 'EnableMemoryMapping');
  }
}

extension on List {
  int get size => length;
}

File: sdk/php/src/Collection.php
<?php
namespace LibraVDB;

require_once __DIR__ . '/LibraVDB.php';

class Collection {
    private int $dbID;
    private int $colID;
    private string $name;
    private int $dimension;

    public function __construct(int $dbID, int $colID, string $name, int $dimension) {
        $this->dbID = $dbID;
        $this->colID = $colID;
        $this->name = $name;
        $this->dimension = $dimension;
    }

    public function insert(string $id, array $vector, string $metadata = "{}"): void {
        if (count($vector) !== $this->dimension) {
            throw new \Exception("Vector dimension mismatch");
        }

        $ffi = LibraVDB::getFFI();
        $c_vector = $ffi->new("float[" . $this->dimension . "]");
        for ($i = 0; $i < $this->dimension; $i++) {
            $c_vector[$i] = $vector[$i];
        }

        $errPtr = $ffi->InsertVector($this->colID, $id, $ffi->cast("float*", $c_vector), $this->dimension, $metadata);
        LibraVDB::checkError($errPtr);
    }

    public function insertBatch(array $ids, array $vectors, ?array $metadata = null): void {
        $count = count($ids);
        if (count($vectors) !== $count) {
            throw new \Exception("Length of ids and vectors must match");
        }

        $ffi = LibraVDB::getFFI();

        // 1. Flatten vectors
        $c_vectors = $ffi->new("float[" . ($count * $this->dimension) . "]");
        for ($i = 0; $i < $count; $i++) {
            if (count($vectors[$i]) !== $this->dimension) {
                throw new \Exception("Vector dimension mismatch at index $i");
            }
            for ($j = 0; $j < $this->dimension; $j++) {
                $c_vectors[$i * $this->dimension + $j] = $vectors[$i][$j];
            }
        }

        // 2. Prepare string arrays
        $c_ids = $ffi->new("const char*[$count]");
        $c_metas = $ffi->new("const char*[$count]");
        
        // We must keep references to FFI strings so they aren't garbage collected
        $strRefs = [];

        for ($i = 0; $i < $count; $i++) {
            // Memory must be allocated or C string conversion handled
            // PHP FFI automatically converts strings when passing them as parameters,
            // but for arrays of strings, we need to manually allocate them.
            // Using FFI::new("char[]") for each string
            
            $idStr = $ids[$i];
            $idLen = strlen($idStr) + 1;
            $cId = $ffi->new("char[$idLen]");
            \FFI::memcpy($cId, $idStr, strlen($idStr));
            $c_ids[$i] = $ffi->cast("const char*", $cId);
            $strRefs[] = $cId;

            $metaStr = $metadata !== null ? $metadata[$i] : "{}";
            $metaLen = strlen($metaStr) + 1;
            $cMeta = $ffi->new("char[$metaLen]");
            \FFI::memcpy($cMeta, $metaStr, strlen($metaStr));
            $c_metas[$i] = $ffi->cast("const char*", $cMeta);
            $strRefs[] = $cMeta;
        }

        $errPtr = $ffi->InsertBatch(
            $this->colID,
            $c_ids,
            $ffi->cast("float*", $c_vectors),
            $count,
            $this->dimension,
            $c_metas
        );

        LibraVDB::checkError($errPtr);
    }

    public function search(array $vector, int $k, string $filter = "{}"): string {
        if (count($vector) !== $this->dimension) {
            throw new \Exception("Vector dimension mismatch");
        }

        $ffi = LibraVDB::getFFI();
        $c_vector = $ffi->new("float[" . $this->dimension . "]");
        for ($i = 0; $i < $this->dimension; $i++) {
            $c_vector[$i] = $vector[$i];
        }

        $resPtr = $ffi->QueryVector($this->colID, $ffi->cast("float*", $c_vector), $this->dimension, $k, $filter);
        return LibraVDB::extractString($resPtr);
    }

    public function scan(int $offset = 0, int $limit = 100): string {
        $ffi = LibraVDB::getFFI();
        $resPtr = $ffi->ScanCollection($this->colID, $offset, $limit);
        return LibraVDB::extractString($resPtr);
    }

    public function delete(string $id): void {
        $ffi = LibraVDB::getFFI();
        $errPtr = $ffi->DeleteVector($this->colID, $id);
        LibraVDB::checkError($errPtr);
    }
}
